File: analytics/src/test/java/io/appmetrica/analytics/impl/TimePassedCheckerTest.java
package io.appmetrica.analytics.impl;

import io.appmetrica.analytics.coreutils.internal.time.TimePassedChecker;
import io.appmetrica.analytics.coreutils.internal.time.TimeProvider;
import io.appmetrica.analytics.testutils.CommonTest;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.when;

public class TimePassedCheckerTest extends CommonTest {

    @Mock
    private TimeProvider mTimeProvider;
    private TimePassedChecker mTimePassedChecker;
    private final long mCurrentTimeSeconds = 12345678;
    private final long mCurrentTimeMillis = 87654321;
    private static final String TAG = "tag";

    @Before
    public void setUp() {
        MockitoAnnotations.openMocks(this);
        when(mTimeProvider.currentTimeSeconds()).thenReturn(mCurrentTimeSeconds);
        when(mTimeProvider.currentTimeMillis()).thenReturn(mCurrentTimeMillis);
        mTimePassedChecker = new TimePassedChecker(mTimeProvider);
    }

    @Test
    public void testCurrentLessThanLastSeconds() {
        assertThat(mTimePassedChecker.didTimePassSeconds(mCurrentTimeSeconds + 1, 0, TAG)).isTrue();
    }

    @Test
    public void testIntervalPassedSeconds() {
        assertThat(mTimePassedChecker.didTimePassSeconds(mCurrentTimeSeconds - 11, 10, TAG)).isTrue();
    }

    @Test
    public void testIntervalNotPassedSeconds() {
        assertThat(mTimePassedChecker.didTimePassSeconds(mCurrentTimeSeconds - 9, 10, TAG)).isFalse();
    }

    @Test
    public void testIntervalPassedExactlySeconds() {
        assertThat(mTimePassedChecker.didTimePassSeconds(mCurrentTimeSeconds - 10, 10, TAG)).isTrue();
    }

    @Test
    public void testCurrentLessThanLastMillis() {
        assertThat(mTimePassedChecker.didTimePassMillis(mCurrentTimeMillis + 1, 0, TAG)).isTrue();
    }

    @Test
    public void testIntervalPassedMillis() {
        assertThat(mTimePassedChecker.didTimePassMillis(mCurrentTimeMillis - 11, 10, TAG)).isTrue();
    }

    @Test
    public void testIntervalNotPassedMillis() {
        assertThat(mTimePassedChecker.didTimePassMillis(mCurrentTimeMillis - 9, 10, TAG)).isFalse();
    }

    @Test
    public void testIntervalPassedExactlyMillis() {
        assertThat(mTimePassedChecker.didTimePassMillis(mCurrentTimeMillis - 10, 10, TAG)).isTrue();
    }
}

File: core-utils/src/test/java/io/appmetrica/analytics/coreutils/internal/time/TimePassedCheckerTest.kt
package io.appmetrica.analytics.coreutils.internal.time

import io.appmetrica.analytics.testutils.CommonTest
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class TimePassedCheckerTest : CommonTest() {

    private val tag = "tag"

    private val currentTimeSeconds = 12345678L
    private val currentTimeMillis = 87654321L
    private val timeProvider = mock<TimeProvider> {
        on { this.currentTimeMillis() } doReturn currentTimeMillis
        on { this.currentTimeSeconds() } doReturn currentTimeSeconds
    }
    private val timePassedChecker = TimePassedChecker(timeProvider)

    @Test
    fun currentLessThanLastSeconds() {
        assertThat(timePassedChecker.didTimePassSeconds(currentTimeSeconds + 1, 0, tag)).isTrue
    }

    @Test
    fun intervalPassedSeconds() {
        assertThat(timePassedChecker.didTimePassSeconds(currentTimeSeconds - 11, 10, tag)).isTrue
    }

    @Test
    fun intervalNotPassedSeconds() {
        assertThat(timePassedChecker.didTimePassSeconds(currentTimeSeconds - 9, 10, tag)).isFalse
    }

    @Test
    fun intervalPassedExactlySeconds() {
        assertThat(timePassedChecker.didTimePassSeconds(currentTimeSeconds - 10, 10, tag)).isTrue
    }

    @Test
    fun currentLessThanLastMillis() {
        assertThat(timePassedChecker.didTimePassMillis(currentTimeMillis + 1, 0, tag)).isTrue
    }

    @Test
    fun intervalPassedMillis() {
        assertThat(timePassedChecker.didTimePassMillis(currentTimeMillis - 11, 10, tag)).isTrue
    }

    @Test
    fun intervalNotPassedMillis() {
        assertThat(timePassedChecker.didTimePassMillis(currentTimeMillis - 9, 10, tag)).isFalse
    }

    @Test
    fun intervalPassedExactlyMillis() {
        assertThat(timePassedChecker.didTimePassMillis(currentTimeMillis - 10, 10, tag)).isTrue
    }
}
